commit 7b8504eaabea6c70195baa3c02155c24a6e270c6
Author: Hanny <me@hannylicious.com>
Date:   Fri Oct 4 12:43:56 2024 -0500

    Fixed #35813 -- Made migrations track all changes to unmanaged models.
    
    Previously, only create and delete operations were generated.

diff --git a/tests/migrations/test_autodetector.py b/tests/migrations/test_autodetector.py
index 7a66e500cb..f2bc7e48a1 100644
--- a/tests/migrations/test_autodetector.py
+++ b/tests/migrations/test_autodetector.py
@@ -712,6 +712,72 @@ class AutodetectorTests(BaseAutodetectorTests):
     author_unmanaged = ModelState(
         "testapp", "AuthorUnmanaged", [], {"managed": False}, ("testapp.author",)
     )
+    author_unmanaged_empty = ModelState(
+        "testapp",
+        "Author",
+        [("id", models.AutoField(primary_key=True))],
+        {"managed": False},
+        ("testapp.author",),
+    )
+    author_unmanaged_name_check_constraint = ModelState(
+        "testapp",
+        "Author",
+        [
+            ("id", models.AutoField(primary_key=True)),
+            ("name", models.CharField(max_length=200, default="Ada Lovelace")),
+        ],
+        {
+            "managed": False,
+            "constraints": [
+                models.CheckConstraint(
+                    condition=models.Q(name__contains="Bob"), name="name_contains_bob"
+                )
+            ],
+        },
+        ("testapp.author",),
+    )
+    author_unmanaged_with_book = ModelState(
+        "testapp",
+        "Author",
+        [
+            ("id", models.AutoField(primary_key=True)),
+            ("name", models.CharField(max_length=200)),
+            ("book", models.ForeignKey("otherapp.Book", models.CASCADE)),
+        ],
+        {"managed": False},
+        ("testapp.author",),
+    )
+    author_unmanaged_name_default = ModelState(
+        "testapp",
+        "Author",
+        [
+            ("id", models.AutoField(primary_key=True)),
+            ("name", models.CharField(max_length=200, default="Ada Lovelace")),
+        ],
+        {"managed": False},
+        ("testapp.author",),
+    )
+    author_unmanaged_name_longer = ModelState(
+        "testapp",
+        "Author",
+        [
+            ("id", models.AutoField(primary_key=True)),
+            ("name", models.CharField(max_length=400)),
+        ],
+        {"managed": False},
+        ("testapp.author",),
+    )
+    author_unmanaged_renamed_with_book = ModelState(
+        "testapp",
+        "Writer",
+        [
+            ("id", models.AutoField(primary_key=True)),
+            ("name", models.CharField(max_length=200)),
+            ("book", models.ForeignKey("otherapp.Book", models.CASCADE)),
+        ],
+        {"managed": False},
+        ("testapp.author",),
+    )
     author_unmanaged_managed = ModelState(
         "testapp", "AuthorUnmanaged", [], {}, ("testapp.author",)
     )
@@ -829,6 +895,17 @@ class AutodetectorTests(BaseAutodetectorTests):
         ],
         {"db_table": "author_three"},
     )
+    book_unmanaged = ModelState(
+        "otherapp",
+        "Book",
+        [
+            ("id", models.AutoField(primary_key=True)),
+            ("author", models.ForeignKey("testapp.Author", models.CASCADE)),
+            ("title", models.CharField(max_length=200)),
+        ],
+        {"managed": False},
+        ("otherapp.book",),
+    )
     contract = ModelState(
         "testapp",
         "Contract",
@@ -968,6 +1045,17 @@ class AutodetectorTests(BaseAutodetectorTests):
             ("title", models.CharField(max_length=200)),
         ],
     )
+    book_with_author_unmanaged_renamed = ModelState(
+        "otherapp",
+        "Book",
+        [
+            ("id", models.AutoField(primary_key=True)),
+            ("author", models.ForeignKey("testapp.Writer", models.CASCADE)),
+            ("title", models.CharField(max_length=200)),
+        ],
+        {"managed": False},
+        ("otherapp.book",),
+    )
     book_with_field_and_author_renamed = ModelState(
         "otherapp",
         "Book",
@@ -4019,8 +4107,45 @@ class AutodetectorTests(BaseAutodetectorTests):
         )
         self.assertEqual(fk_field.remote_field.model, "testapp.AAuthorProxyProxy")
 
+    def test_unmanaged_add_constraints(self):
+        """Test change detection of new constraints on an unmanaged model."""
+        changes = self.get_changes(
+            [self.author_unmanaged_name_default],
+            [self.author_unmanaged_name_check_constraint],
+        )
+        self.assertNumberMigrations(changes, "testapp", 1)
+        self.assertOperationTypes(changes, "testapp", 0, ["AddConstraint"])
+        added_constraint = models.CheckConstraint(
+            condition=models.Q(name__contains="Bob"), name="name_contains_bob"
+        )
+        self.assertOperationAttributes(
+            changes, "testapp", 0, 0, model_name="author", constraint=added_constraint
+        )
+
+    def test_unmanaged_add_field(self):
+        """Tests autodetection of new fields on an unmanaged model."""
+        changes = self.get_changes(
+            [self.author_unmanaged_empty], [self.author_unmanaged_name_default]
+        )
+        # Right number/type of migrations?
+        self.assertNumberMigrations(changes, "testapp", 1)
+        self.assertOperationTypes(changes, "testapp", 0, ["AddField"])
+        self.assertOperationAttributes(changes, "testapp", 0, 0, name="name")
+
+    def test_unmanaged_alter_field(self):
+        """Tests autodetection of altered fields on an unmanaged model."""
+        changes = self.get_changes(
+            [self.author_unmanaged_name_default], [self.author_unmanaged_name_longer]
+        )
+        # Right number/type of migrations?
+        self.assertNumberMigrations(changes, "testapp", 1)
+        self.assertOperationTypes(changes, "testapp", 0, ["AlterField"])
+        self.assertOperationAttributes(
+            changes, "testapp", 0, 0, name="name", preserve_default=True
+        )
+
     def test_unmanaged_create(self):
-        """The autodetector correctly deals with managed models."""
+        """The autodetector correctly deals with unmanaged models."""
         # First, we test adding an unmanaged model
         changes = self.get_changes(
             [self.author_empty], [self.author_empty, self.author_unmanaged]
@@ -4032,6 +4157,50 @@ class AutodetectorTests(BaseAutodetectorTests):
             changes, "testapp", 0, 0, name="AuthorUnmanaged", options={"managed": False}
         )
 
+    def test_unmanaged_remove_field(self):
+        """Tests autodetection of removed fields on an unmanaged model."""
+        changes = self.get_changes(
+            [self.author_unmanaged_name_default], [self.author_unmanaged_empty]
+        )
+        # Right number/type of migrations?
+        self.assertNumberMigrations(changes, "testapp", 1)
+        self.assertOperationTypes(changes, "testapp", 0, ["RemoveField"])
+        self.assertOperationAttributes(
+            changes, "testapp", 0, 0, name="name", model_name="author"
+        )
+
+    def test_unmanaged_remove_constraints(self):
+        """Test change detection of new constraints on an umnanaged model."""
+        changes = self.get_changes(
+            [self.author_unmanaged_name_check_constraint],
+            [self.author_unmanaged_name_default],
+        )
+        self.assertNumberMigrations(changes, "testapp", 1)
+        self.assertOperationTypes(changes, "testapp", 0, ["RemoveConstraint"])
+        self.assertOperationAttributes(
+            changes, "testapp", 0, 0, model_name="author", name="name_contains_bob"
+        )
+
+    def test_unmanaged_rename_model(self):
+        """Tests autodetection of renamed models on an unmanaged model."""
+        changes = self.get_changes(
+            [self.author_unmanaged_with_book, self.book_unmanaged],
+            [
+                self.author_unmanaged_renamed_with_book,
+                self.book_with_author_unmanaged_renamed,
+            ],
+            MigrationQuestioner({"ask_rename_model": True}),
+        )
+        # Right number/type of migrations?
+        self.assertNumberMigrations(changes, "testapp", 1)
+        self.assertOperationTypes(changes, "testapp", 0, ["RenameModel"])
+        self.assertOperationAttributes(
+            changes, "testapp", 0, 0, old_name="Author", new_name="Writer"
+        )
+        # Now that RenameModel handles related fields too, there should be
+        # no AlterField for the related field.
+        self.assertNumberMigrations(changes, "otherapp", 0)
+
     def test_unmanaged_delete(self):
         changes = self.get_changes(
             [self.author_empty, self.author_unmanaged], [self.author_empty]
